pkg/util/kubernetes/kubelet/kubelet_client.go

```diff
@@ -0,0 +1,301 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+// +build kubelet
+
+package kubelet
+
+import (
+	"crypto/tls"
+	"errors"
+	"expvar"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util/filesystem"
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+var (
+	kubeletExpVar = expvar.NewInt("kubeletQueries")
+)
+
+type kubeletClientConfig struct {
+	scheme         string
+	baseURL        string
+	tlsVerify      bool
+	caPath         string
+	clientCertPath string
+	clientKeyPath  string
+	token          string
+}
+
+type kubeletClient struct {
+	client     http.Client
+	kubeletURL string
+	headers    http.Header
+	config     kubeletClientConfig
+}
+
+func newForConfig(config kubeletClientConfig, timeout time.Duration) (*kubeletClient, error) {
+	var err error
+
+	// Building transport based on options
+	customTransport := http.DefaultTransport.(*http.Transport).Clone()
+
+	// Building custom TLS config
+	tlsConfig := &tls.Config{}
+	tlsConfig.InsecureSkipVerify = !config.tlsVerify
+
+	if config.caPath == "" && filesystem.FileExists(kubernetes.DefaultServiceAccountCAPath) {
+		config.caPath = kubernetes.DefaultServiceAccountCAPath
+	}
+
+	if config.caPath != "" {
+		tlsConfig.RootCAs, err = kubernetes.GetCertificateAuthority(config.caPath)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if config.clientCertPath != "" && config.clientKeyPath != "" {
+		tlsConfig.Certificates, err = kubernetes.GetCertificates(config.clientCertPath, config.clientKeyPath)
+		if err != nil {
+			return nil, err
+		}
+	}
+	customTransport.TLSClientConfig = tlsConfig
+
+	// Do not use token in plain text
+	headers := http.Header{}
+	if config.scheme == "https" {
+		if config.token != "" {
+			headers.Set(authorizationHeaderKey, fmt.Sprintf("bearer %s", config.token))
+		}
+	}
+
+	// Defaulting timeout
+	if timeout == 0 {
+		timeout = 30 * time.Second
+	}
+
+	return &kubeletClient{
+		client: http.Client{
+			Transport: customTransport,
+			Timeout:   timeout,
+		},
+		kubeletURL: fmt.Sprintf("%s://%s", config.scheme, config.baseURL),
```

Comment:
Have you looked at url.Join or path.Join?

Reply:
Apparently `url.Join` does not exist and `path.Join` does not handle URL directly, so I think it's more straightforward this way.